### ISSUE TITLE
Identity: TwoFactorRememberMe scheme should be signed out if authorization for this scheme failed due to the failure of validation of the security stamp (leads to login/logout loop if the next login does not sign in two factor remember this machine).

### ISSUE THREAD

**Issue by rwasef1830** [dotnet/aspnetcore#37963]
Reproduced on .NET 5.

Steps:
1. Setup two factor authentication.
2. Log out, login with two factor, remember the machine.
3. Change the user's security stamp.
4. Try to authenticate the TwoFactorRememberMe scheme, the security stamp validation will fail and you will get logged out.
5. Login using a method that does not sign in again to the TwoFactorRememberMe scheme (eg: external login).
6. Invoke any code path that tries to authenticate the TwoFactorRememberMe scheme, the security stamp validation will fail and you will get logged out. Go to step 5. (infinite loop).

Cause of issue:
- When the user's security stamp gets updated for any reason, all sign in schemes need to get refreshed, if for some reason the two factor remember me ticket security stamp goes out of sync, when the validation fails, the two factor remember me cookie also needs to be cleared. 
- In the default templates, visiting any page which calls `SignInManager.IsTwoFactorClientRememberedAsync` will log out the user, but since this logging out does not log out the TwoFactorRememberMe scheme, the invalid state is never resolved until the user explicitly logs in this scheme to reset it (which may not happen if user uses an external login).

Will make a repro repo as soon as I get some time.

Workaround:
- Hook into the security stamp validation and when the validation fails for the two factor remember me, sign it out also (along with the other schemes).

**Comment by rwasef1830** [comment]
The same issue is also present for TwoFactorUserId scheme.

**Comment by HaoK** [comment]
So to summarize, it sounds like you have some code paths that don't use 2fa, and if you get into a state where you have a bad 2fa remember me set, and try a code path that doesn't require 2fa you might get stuck in an infinite loop since the 2fa remember me will never be clear and is never going to be valid?

A repro would definitely help isolate if any change is needed in the templates

**Comment by HaoK** [comment]
Also do you have bypass twofactor set on your external logins?

**Comment by rwasef1830** [comment]
@HaoK that's correct. I have it so that external login does not sign in two factor and doesn't require it (since I will delegate the 2 factor burden to the external service).

I worked around it in my code by inheriting the SignInManager and implementing the following 2 overrides:

```c#
        // https://github.com/dotnet/aspnetcore/issues/37963
        public override async Task<TUser> ValidateTwoFactorSecurityStampAsync(ClaimsPrincipal principal)
        {
            var user = await base.ValidateTwoFactorSecurityStampAsync(principal).ConfigureAwait(false);
            if (user == null)
            {
                await this.Context.SignOutAsync(IdentityConstants.TwoFactorUserIdScheme).ConfigureAwait(false);
                await this.Context.SignOutAsync(IdentityConstants.TwoFactorRememberMeScheme).ConfigureAwait(false);
            }

            return user!;
        }

        public override async Task RefreshSignInAsync(TUser user)
        {
            await base.RefreshSignInAsync(user).ConfigureAwait(false);
            await this.Context.SignOutAsync(IdentityConstants.TwoFactorUserIdScheme).ConfigureAwait(false);
            await this.Context.SignOutAsync(IdentityConstants.TwoFactorRememberMeScheme).ConfigureAwait(false);
        }
```

I think this needs to be in the framework level, not just in the templates, since it's a logical mistake in my opinion.

**Comment by HaoK** [comment]
So for more clarification, everything will work fine even after security stamp has been reset if you go through the normal 2fa enabled flows, it's just that when you try to go through external sign in, what's blocking you exactly?  If you are passing bypassTwoFactor = true to https://github.com/dotnet/aspnetcore/blob/main/src/Identity/Core/src/SignInManager.cs#L616 you should be getting the remember check skipped entirely no? https://github.com/dotnet/aspnetcore/blob/6325163f926cb763860a2a92beec0e20ca3daae1/src/Identity/Core/src/SignInManager.cs#L783

**Comment by rwasef1830** [comment]
@HaoK if you change the security stamp, and then logout, then login with external sign in bypassing 2fa, then as soon as you go view the user's 2 factor settings page, you will get logged out because the validation of the security stamp failed for the now outdated previously set two factor remember me cookie.

**Comment by HaoK** [comment]
So I wasn't able to reproduce this using the following steps on a 6.0 template where I added facebook auth for the external sign in that bypasses 2fa.

`dotnet new mvc --auth Individual`
add facebook external authentication to the app
register a user via facebook, confirm account, login, enable 2fa via authenticator app, add a local password, logout
login via password and enable remember this machine via 2fa dialog
logout and then login via facebook, go to 2fa page under manage, this all worked for me.

What's the flow that demonstrates the behavior you are describing?

**Comment by rwasef1830** [comment]
That exact flow, are you still logged in if you refresh the page after the 2fa page loads once ? (the signout happens during the loading of the page in the IsTwoFactorClientRememberedAsync check).

**Comment by HaoK** [comment]
I'm still logged in, I just tried with a 1 second validation interval, I don't get logged out due to refresh sign in on the change password page, are you triggering the security stamp change via some other mechanism?

**Comment by rwasef1830** [comment]
@HaoK 
Change the password to invalidate the "remember this machine" security stamp before you log out to trigger the issue.

> dotnet new mvc --auth Individual
add facebook external authentication to the app
register a user via facebook, confirm account, login, enable 2fa via authenticator app, add a local password, logout
login via password and enable remember this machine via 2fa dialog
**Change the password**
logout and then login via facebook, go to 2fa page under manage, refresh the page -> Signed out

**Comment by HaoK** [comment]
Yeah I did that as well, with the short 1s validation window it actually logs me out before I can get to the 2fa page, so I have to relogin again.

Anyways, the security stamp validator is only something that's tied to the main application cookie, it should have no effect on the other two cookies as those should stick around irrespective of other settings as its a separate cookie, so the behavior I'm seeing is what I would expect, you can change your password, login via password or facebook, and 2fa is skipped because that cookie is still there.

I confirmed that via F12 the Identity.TwoFactorRememberMe cookie does not get lost.

**Comment by rwasef1830** [comment]
The two factor remember me cookie indeed does not get lost, the problem is that it's not refreshed when the user's security stamp is changed, and then any code path that triggers the revalidation of this cookie (eg: the SignInManager.IsTwoFactorClientRememberdAsync call) will cause the user to get kicked out of the login.

The TwoFactorRememberMe scheme cookie does have a security stamp validator, in fact it's here:
https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Identity/Core/src/IdentityServiceCollectionExtensions.cs#L64

And it calls into here:
https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Identity/Core/src/SignInManager.cs#L293

And when that call returns a null user because the user's current security stamp no longer matches the one at the time that cookie was issued, this code signs the user out:
https://github.com/dotnet/aspnetcore/blob/4736d1c4953d73b1db9e01cc919001808144cff1/src/Identity/Core/src/SecurityStampValidator.cs#L144

The TwoFactor security stamp validator inherits from the default security stamp validator, so when its ValidateAsync call fails, it calls SignInManager.SignOutAsync which then signs out all schemes except the two factor remember me (which also needs to be signed out since it now has a wrong security stamp):
https://github.com/dotnet/aspnetcore/blob/4736d1c4953d73b1db9e01cc919001808144cff1/src/Identity/Core/src/SignInManager.cs#L256

Normally when a user asks to Sign out we need to keep the two factor remember me cookie, *but in case the signout was triggered by the failure of validation of the security stamp of the two factor remember me cookie itself, the two factor remember scheme needs to be signed out of as well*.

The user ends up having a two factor remember me with a stale security stamp that never gets cleared unless they explicitly login with a normal password and explicitly opt to remember the machine to set a new two factor remember me cookie with the correct security stamp.

I'm a bit short of time, I will put up a repro page that clearly shows the issue I am talking about in some hours. Thank you for your patience with me.

**Comment by HaoK** [comment]
Okay, I don't think we need a full repro, I understand the issue now, thanks for debugging this so thoroughly, I forgot that we added security stamp validation to the 2fa (https://github.com/dotnet/aspnetcore/commit/c6a82ad19a0bdb80b7358dc843b66664bbf33f96)

I think we could fix this with just an overload of SignOutAsync which has an additional bool signOutTwoFactorRememberMe which we can call it set to true in only this case.  Would that also have worked in your scenario?  Seems reasonable for RefreshSignIn to also refresh the 2faRememberMe as well, since the whole idea of this method is to preserve all the old state to prevent the user from having to relog in/do 2fa after doing an intentional action that changes their stamp

**Comment by rwasef1830** [comment]
You need to sign out the two factor remember me in case the sign out was triggered from inside the security stamp validation. Whereas a normal sign out should preserve the remember me cookie. So the security stamp validation stamp code path would need to call this new overload. It would also constitute a breaking change to the SignOutAsync API.

Refreshing the sign in is tricky. The way the API is structured, you can't check if a scheme is "authenticated but without validation" to decide whether to refresh it or not, and you don't want to accidentally sign in with the two factor remember me if the user did not opt for that. The only safe action is to sign it out in my opinion on security stamp changes. The other alternative would be to make a way to check the existence of a "present but invalid" TwoFactorRememberMe signed in ticket but I think that would be surgery in the API and too many breaking changes.

**Comment by rwasef1830** [comment]
@HaoK the reason you didn't see the issue in the default templates is that AddDefaultIdentity used in the templates does **not** setup security stamp validation for the 2FA. Only the "full" AddIdentity does.

**Comment by rwasef1830** [comment]
My workaround is basically what you are suggesting, to answer your question. see my comment above, it's kind of a hack but works.

**Comment by HaoK** [comment]
PR has been merged for 7, need to create backport for 6.0

**Comment by petursig** [comment]
> next

A workaround for this would be to check if the user has 2fa enabled before calling IsTwoFactorClientRemebered, like so:

            Is2faEnabled = await _userManager.GetTwoFactorEnabledAsync(user);
            if(Is2faEnabled)
                IsMachineRemembered = await _signInManager.IsTwoFactorClientRememberedAsync(user);
                
This should be updated in the scaffolded Identity pages.
